structured_prediction_baselines/modules/sampler/sequence_tagging/sequence_tagging_inference.py

```diff
@@ -0,0 +1,224 @@
+        self._different_training_and_eval = True
+
+    @classmethod
+    def from_partial_objects(
+        cls,
+        optimizer: Lazy[Optimizer],
+        loss_fn: Lazy[Loss],  #: This loss can be different from the main loss
+        inference_nn: Optional[TaskNN] = None,
+        cost_augmented_layer: Optional[torch.nn.Module] = None,
+        score_nn: Optional[ScoreNN] = None,
+        oracle_value_function: Optional[OracleValueFunction] = None,
+        stopping_criteria: Union[int, StoppingCriteria] = 1,
+    ) -> "SequenceTaggingInference":
+        loss_fn_ = loss_fn.construct(
+            score_nn=score_nn, oracle_value_function=oracle_value_function
+        )
+        trainable_parameters = {}
+        trainable_parameters.update(inference_nn.named_parameters())
+        trainable_parameters.update(cost_augmented_layer.named_parameters())
+        optimizer_ = optimizer.construct(model_parameters=list(trainable_parameters.items()))
+
+        return cls(
+            optimizer_,
+            loss_fn_,
+            inference_nn=inference_nn,
+            score_nn=score_nn,
+            cost_augmented_layer=cost_augmented_layer,
+            oracle_value_function=oracle_value_function,
+            stopping_criteria=stopping_criteria,
+        )
+
+    def get_loss_fn(
+        self, x: Any, labels: Optional[torch.Tensor]
+    ) -> Callable[[torch.Tensor], torch.Tensor]:
+        # Sampler gets labels of shape (batch, ...), hence this
+        # function will get labels of shape (batch*num_init_samples, ...)
+        # but Loss expect y or shape (batch, num_samples or 1, ...)
+        # Also during eval the loss is different. We inform the loss function
+        # using None
+
+        if self.training and (labels is None):
+            warnings.warn("Labels should not be None in training mode!")
+
+        def loss_fn(y_hat: torch.Tensor, y_cost_aug: torch.Tensor, buffer: Dict) -> torch.Tensor:
+            return self.loss_fn(
+                x,
+                (
+                    labels  # E:labels.unsqueeze(1)
+                    if (self.training and labels is not None)
+                    else None
+                ),
+                y_hat,
+                y_cost_aug,
+                buffer,
+                None,
+            )
+
+        return loss_fn
+
+    def get_dtype_device(self) -> Tuple[torch.dtype, torch.device]:
+        for param in self.loss_fn.parameters():
+            dtype = param.dtype
+            device = param.device
+
+            break
+
+        return dtype, device
+
+    def get_batch_size(self, x: Any) -> int:
+        if isinstance(x, torch.Tensor):
+            return x.shape[0]
+        else:
+            raise NotImplementedError
+
+    @contextlib.contextmanager
+    def param_grad(self) -> Generator[None, None, None]:
+        trainable_modules = [self.inference_nn, self.cost_augmented_layer]
+        requires_grad_map = {}
+        for module in trainable_modules:
```

Comment:
I think this will not work. This for loop should be repeated both in try and finally.